File: Omni/Shared/AppState.swift
import SwiftUI
import Combine

class AppState: ObservableObject {
    @Published var currentUser: UserProfile?
    @Published var currentRoute: AppRoute = .onboarding
    
    // Simple persistence simulation (UserDefaults or just memory for MVP session)
    // For MVP validation, let's keep it in memory mostly, but maybe save to UserDefaults to skip onboarding on relaunch if needed.
    // For now, let's assume fresh start or simple persistence.
    
    init() {
        // Load user from disk if exists (Mock implementation)
        // self.currentUser = loadUser()
        // if self.currentUser != nil { currentRoute = .home }
    }
    
    func completeOnboarding(dob: Date, petName: String, petType: String, petImage: UIImage?) {
        // 1. Generate DNA (Mock Logic)
        let dna = generateMockDNA(dob: dob)
        let soul = generateMockSoulID(petType: petType)
        
        // 2. Create User
        let user = UserProfile(
            dateOfBirth: dob,
            petName: petName,
            petType: petType,
            petImage: petImage?.jpegData(compressionQuality: 0.8),
            energyDNA: dna,
            soulID: soul
        )
        
        // 3. Save & Transition
        self.currentUser = user
        withAnimation {
            self.currentRoute = .home
        }
    }
    
    private func generateMockDNA(dob: Date) -> EnergyDNA {
        // Simple logic: Odd/Even day -> Different types
        let calendar = Calendar.current
        let day = calendar.component(.day, from: dob)
        
        if day % 2 == 0 {
            return EnergyDNA(type: "Void Walker", description: "You thrive in chaos and silence.", colorHex: "#4A4E69")
        } else {
            return EnergyDNA(type: "Solar Flare", description: "You burn bright and exhaust quickly.", colorHex: "#FFB84C")
        }
    }
    
    private func generateMockSoulID(petType: String) -> SoulID {
        let suffix = Int.random(in: 1000...9999)
        let archetype = petType.lowercased() == "cat" ? "Master Manipulator" : "Loyal Goofball"
        return SoulID(id: "OMNI-\(suffix)", archetype: archetype, quote: "Judge me all you want, I know where you sleep.")
    }
}

File: omni_flutter/lib/providers/app_state.dart
import 'dart:math';
import 'dart:typed_data';
import 'package:flutter/material.dart';
import '../models/user_profile.dart';
import '../services/memory_service.dart';

class AppState extends ChangeNotifier {
  UserProfile? _currentUser;
  bool _isOnboarding = true;
  final MemoryService _memoryService = MemoryService();

  UserProfile? get currentUser => _currentUser;
  bool get isOnboarding => _isOnboarding;

  AppState() {
    _loadUserData();
  }

  // Load user data from memory on app start
  Future<void> _loadUserData() async {
    final profile = await _memoryService.getUserProfile();
    if (profile != null) {
      _isOnboarding = false;
      _currentUser = UserProfile(
        dateOfBirth: DateTime.parse(profile['birthday']),
        petName: profile['petName'],
        petType: profile['petType'],
        petImage: null,
        energyDNA: EnergyDNA(
          type: profile['energyDNA'] ?? 'Unknown',
          description: profile['energyDescription'] ?? '',
          colorHex: int.tryParse(profile['colorHex'] ?? '0xFF4A4E69') ?? 0xFF4A4E69,
        ),
        soulID: SoulID(
          id: profile['soulID'] ?? 'OMNI-0000',
          archetype: profile['archetype'] ?? 'Unknown',
          quote: profile['quote'] ?? '',
        ),
      );
      notifyListeners();
    }
  }

  void completeOnboarding({
    required DateTime dateOfBirth,
    required String petName,
    required String petType,
    Uint8List? petImage,
  }) async {
    final dna = _generateMockDNA(dateOfBirth);
    final soul = _generateMockSoulID(petType);

    _currentUser = UserProfile(
      dateOfBirth: dateOfBirth,
      petName: petName,
      petType: petType,
      petImage: petImage,
      energyDNA: dna,
      soulID: soul,
    );

    _isOnboarding = false;

    // Save to memory
    await _memoryService.saveUserProfile({
      'birthday': dateOfBirth.toIso8601String(),
      'petName': petName,
      'petType': petType,
      'energyDNA': dna.type,
      'energyDescription': dna.description,
      'colorHex': dna.colorHex.toString(),
      'soulID': soul.id,
      'archetype': soul.archetype,
      'quote': soul.quote,
    });

    notifyListeners();
  }

  EnergyDNA _generateMockDNA(DateTime dob) {
    final day = dob.day;

    if (day % 2 == 0) {
      return EnergyDNA(
        type: 'Void Walker',
        description: 'You thrive in chaos and silence.',
        colorHex: 0xFF4A4E69,
      );
    } else {
      return EnergyDNA(
        type: 'Solar Flare',
        description: 'You burn bright and exhaust quickly.',
        colorHex: 0xFFFFB84C,
      );
    }
  }

  SoulID _generateMockSoulID(String petType) {
    final random = Random();
    final suffix = 1000 + random.nextInt(9000);
    final archetype = petType.toLowerCase() == 'cat'
        ? 'Master Manipulator'
        : 'Loyal Goofball';

    return SoulID(
      id: 'OMNI-$suffix',
      archetype: archetype,
      quote: 'Judge me all you want, I know where you sleep.',
    );
  }
}
